refactor(core): static-query schematic should handle binary expressions and function callbacks

devversion opened the pull request:
* See individual commits.

devversion committed:
refactor(core): static-query schematic should handle binary expressions

Currently we only check getters for property access expressions. This is wrong
because property access expressions do not always cause the "getter" to be
triggered. e.g.

```ts
set a() {...}
get a() {...}

ngOnInit() {
  this.a = true;
}
```

In that case the schematic currently incorrectly checks the "getter", while this is a binary
expression and the property access is used as left-side of the binary expression. In that
case we need to check the setter declaration of the property and not the "getter". In order
to fix this, we need to also check `ts.BinaryExpression` nodes and check getters/setters
based on the used operator token. There are three types of binary expressions:

1) Value assignment (using `=`). In that case only the setter is triggered.
2) Compound assignment (e.g. using `+=`). In that case `getter` and `setter` are triggered.
3) Comparison (e.g. using `===`). In that case only the getter is triggered.

kara reviewed (approved):
LGTM, aside from duplicated test. Open question can be resolved in a follow-up.

devversion commented on packages/core/schematics/test/static_queries_migration_spec.ts:
```diff
@@ -639,6 +639,121 @@ describe('static-queries migration', () => {
+          .toContain(`@${queryType}('test', { static: true }) query: any;`);
+    });
+
+    it('should detect callback with statically used queries', () => {
```
Good catch. :+1:

devversion commented on packages/core/schematics/test/static_queries_migration_spec.ts:
```diff
@@ -639,6 +639,121 @@ describe('static-queries migration', () => {
+
+      expect(tree.readContent('/index.ts'))
+          .toContain(`@${queryType}('test', { static: true }) query: any;`);
+    });
```
Hmm that is a good point. The initial thinking of the usage detection was that we just ignore the conditionals as we can't know to what the evaluate. So we always mark the query as `static` in that case.

Marking the query as _ambiguous_ with `static: true` sounds reasonable though. As you said, let's do/ think about this more in a follow-up. There are plenty other things that we discussed about. I added it to the query detection strategy hackmd.

devversion committed:
refactor(core): static-query schematic should handle function callbacks

Currently the static-query schematic is not able to properly handle
call expressions that pass function declarations that access a given
query. e.g.

```ts
ngOnInit() {
  this._callFunction(() => this.myQuery.doSomething());
}

_callFunction(cb: any) { cb(); }
```

In that case the passed function is executed synchronously in
the "ngOnInit" lifecycle and therefore the query needs to be
detected as "static".

We can fix this by keeping track of the current function context
and using it to resolve identifiers to the passed arguments.

kara commented:
[presubmit](http://test/OCL:242510667:BASE:242510708:1554750575196:b31495f5)

kara commented:
merge-assistance: I don't know what's going on with the "cla/google" check, but it's ready to go
